File: izzu/izzu/report/supplier_statement/supplier_statement.js
// Copyright (c) 2018, Frappe Technologies Pvt. Ltd. and Contributors
// License: GNU General Public License v3. See license.txt

frappe.query_reports["Supplier Statement"] = {
    filters: [
        {
            fieldname: "party_type",
            label: __("Party Type"),
            fieldtype: "Select",
            options: [
                { "value": "Supplier", "label": __("Supplier") }
            ],
            default: "Supplier",
            hidden: 1,
            // on_change: function () {
            //     frappe.query_report.set_filter_value("party", "");
            // }
        },
        {
            fieldname: "party",
            label: __("Supplier"),
            fieldtype: "Link",
            options: "Supplier",
            // get_data: function (txt) {
            //     return frappe.db.get_link_options("Supplier", txt);
            // },

            on_change: function () {
                let parties = frappe.query_report.get_filter_value("party");
                // if (parties.length === 0 || parties.length > 1) {
                //     console.log(parties);
                    
                //     frappe.query_report.set_filter_value("party_name", "");
                //     // frappe.query_report.set_filter_value("tax_id", "");
                // } else {
                //     let party = parties[0];
                //     frappe.db.get_value("Supplier", party, ["name", "tax_id"], function (value) {
                //         frappe.query_report.set_filter_value("party_name", value.name);
                //         // frappe.query_report.set_filter_value("tax_id", value.tax_id);
                //     });

                    // let supplier = frappe.query_report.get_filter_value('supplier');
                    if (parties) {
                        frappe.db.get_doc('Supplier', parties).then(supplier_doc => {
                            localStorage.setItem("supplier", JSON.stringify({
                                id: supplier_doc.name,
                                name: supplier_doc.supplier_name,
                            }));

                            if (supplier_doc.default_currency) {
                                localStorage.setItem("supplier_currency", JSON.stringify({
                                    currency: supplier_doc.default_currency,
                                }));
                            } else {
                                let company = frappe.query_report.get_filter_value('company');
                                if (company) {
                                    frappe.db.get_doc('Company', company).then(company_doc => {
                                        localStorage.setItem("supplier_currency", JSON.stringify({
                                            currency: company_doc.default_currency,
                                        }));
                                    });
                                }
                            }

                            if (supplier_doc.supplier_primary_address) {
                                frappe.db.get_doc('Address', supplier_doc.supplier_primary_address)
                                    .then(address_doc => {
                                        let address_parts = [
                                            address_doc.address_line1 || "",
                                            address_doc.address_line2 || "",
                                            address_doc.city || "",
                                            address_doc.country || ""
                                        ];
                                        let full_address = address_parts.filter(Boolean).join('<br>');
                                        localStorage.setItem("supplier_address", JSON.stringify({
                                            address: full_address
                                        }));
                                    });
                            } else {
                                localStorage.setItem("supplier_address", JSON.stringify({
                                    address: ''
                                }));
                            }

                            if (supplier_doc.supplier_primary_contact) {
                                frappe.db.get_doc('Contact', supplier_doc.supplier_primary_contact)
                                    .then(contact_doc => {
                                        let phone_no = contact_doc.phone_nos?.find(row => row.is_primary_phone)?.phone || "";
                                        let email_id = contact_doc.email_ids?.find(row => row.is_primary)?.email_id || "";
                                        setTimeout(() => {
                                            localStorage.setItem("supplier_contact", JSON.stringify({
                                                phone: phone_no,
                                                email: email_id
                                            }));
                                        }, 500);
                                    });
                            } else {
                                setTimeout(() => {
                                    localStorage.setItem("supplier_contact", JSON.stringify({
                                        phone: '',
                                        email: ''
                                    }));
                                }, 500);
                            }
                        });
                    } else {
                        localStorage.removeItem("supplier");
                        localStorage.removeItem("supplier_address");
                        localStorage.removeItem("supplier_contact");
                    }
                }
            
        },
        {
            fieldname: "from_date",
            label: __("From Date"),
            fieldtype: "Date",
            width: 100,
            default: frappe.datetime.add_months(frappe.datetime.get_today(), -1),
            on_change: function(report) {
                let from_date = frappe.query_report.get_filter_value('from_date');
                localStorage.setItem("from_date", JSON.stringify({ from_date: from_date || {} }));
                report.refresh();
            }
        },
        {
            fieldname: "to_date",
            label: __("To Date"),
            fieldtype: "Date",
            width: 100,
            default: frappe.datetime.get_today(),
            on_change: function(report) {
                let to_date = frappe.query_report.get_filter_value('to_date');
                localStorage.setItem("to_date", JSON.stringify({ to_date: to_date || {} }));
                report.refresh();
            }
        },
        {
            fieldname: "company",
            label: __("Company"),
            fieldtype: "Link",
            options: "Company",
            width: 100,
            reqd: 1,
            default: frappe.defaults.get_user_default("Company"),
            on_change: function(report) {
                let company = frappe.query_report.get_filter_value('company');
                if (company) {
                    frappe.db.get_list('Dynamic Link', {
                        filters: {
                            link_doctype: "Company",
                            link_name: company,
                            link_title: company
                        },
                        fields: ['parent']
                    }).then(records => {
                        let office_address = "";
                        if (records?.length) {
                            frappe.db.get_doc('Address', records[0].parent).then(address_doc => {
                                if (address_doc.address_type === "Office") {
                                    let address_parts = [
                                        address_doc.address_line1 || "",
                                        address_doc.address_line2 || "",
                                        address_doc.city || "",
                                        address_doc.country || ""
                                    ];
                                    office_address = address_parts.filter(Boolean).join('<br>');
                                    localStorage.setItem("company_address", JSON.stringify({
                                        address: office_address
                                    }));
                                }
                            });
                        }
                    });

                    frappe.db.get_doc('Company', company).then(company_doc => {
                        localStorage.setItem("company", JSON.stringify({
                            name: company_doc.name,
                            phone: company_doc.phone_no,
                            email: company_doc.email,
                            website: company_doc.website,
                        }));
                    });
                } else {
                    localStorage.setItem("company", JSON.stringify({}));
                    localStorage.setItem("company_address", JSON.stringify({}));
                }
                report.refresh();
            }
        },
        {
            fieldname: "group_by",
            label: __("Group by"),
            fieldtype: "Select",
            options: [
                "",
                { label: __("Group by Voucher"), value: "Group by Voucher" },
                { label: __("Group by Voucher (Consolidated)"), value: "Group by Voucher (Consolidated)" },
                { label: __("Group by Account"), value: "Group by Account" },
                { label: __("Group by Party"), value: "Group by Party" },
            ],
            default: "Group by Voucher (Consolidated)",
            hidden: 1
        },
    ],
    onload: function (report) {
        let to_date = frappe.query_report.get_filter_value('to_date');
        localStorage.setItem("to_date", JSON.stringify({ to_date: to_date || {} }));

        setTimeout(() => {
            localStorage.setItem("current_date", new Date().toISOString().split('T')[0]);
            frappe.db.get_doc('User', frappe.session.user).then(doc => {
                localStorage.setItem("session_user", doc.full_name);
            });
            $('[data-fieldname="vehicle"]').hide();
        }, 100);
        report.refresh();
    },
};

erpnext.utils.add_dimensions("Supplier Statement", 15);
 

File: izzu/izzu/report/supplier_statement/supplier_statement.py
import frappe
from frappe.utils import flt

def execute(filters=None):
    if not filters:
        filters = {}

    from_date = filters.get("from_date")
    to_date = filters.get("to_date")
    supplier_filter = filters.get("supplier")

    # Get opening balances
    opening_balances = get_opening_balances(from_date, supplier_filter)

    # Get transactions
    transactions = get_transactions(from_date, to_date, supplier_filter)

    # Combine data
    data = []
    suppliers = set(opening_balances.keys()) | set(transactions.keys())

    for supplier in sorted(suppliers):
        opening = opening_balances.get(supplier, 0)
        txns = transactions.get(supplier, [])

        # Fetch supplier name
        supplier_name = frappe.db.get_value("Supplier", supplier, "supplier_name") or supplier

        if txns or opening != 0:
            data.append({
                "date": "",
                "supplier": supplier,
                "supplier_name": supplier_name,
                "voucher_subtype": "",
                "voucher_type": "",
                "voucher_no": "",
                "doc_ref": "<b>Opening Balance</b>",
                "opening_balance": opening,
                "debit": 0,
                "credit": 0,
                "balance": opening
            })

        total_debit = total_credit = 0
        balance = opening

        for txn in txns:
            balance += flt(txn["debit"]) - flt(txn["credit"])
            data.append({
                "date": txn.get("posting_date"),
                "supplier": supplier,
                "supplier_name": "",
                "voucher_subtype": txn["voucher_subtype"],
                "voucher_type": txn["voucher_type"],
                "voucher_no": txn["voucher_no"],
                "doc_ref": txn.get("doc_ref", ""),
                "opening_balance": "",
                "debit": txn["debit"],
                "credit": txn["credit"],
                "balance": balance
            })
            total_debit += flt(txn["debit"])
            total_credit += flt(txn["credit"])

        if txns or opening != 0:
            data.append({
                "date": "",
                "supplier": "",
                "supplier_name": "",
                "voucher_subtype": "",
                "voucher_type": "",
                "voucher_no": "",
                "doc_ref": "<b>Total</b>",
                "opening_balance": "",
                "debit": total_debit,
                "credit": total_credit,
                "balance": balance
            })
            data.append({"page_break": 0})

    return get_columns(), data

def get_opening_balances(from_date, supplier_filter):
    condition = "posting_date < %(from_date)s AND party_type = 'Supplier'"
    params = {"from_date": from_date}

    if supplier_filter:
        condition += " AND party = %(supplier)s"
        params["supplier"] = supplier_filter

    result = frappe.db.sql(f"""
        SELECT party, SUM(debit - credit) AS opening_balance
        FROM `tabGL Entry`
        WHERE {condition}
        GROUP BY party
    """, params, as_dict=1)

    return {row.party: flt(row.opening_balance) for row in result}

def get_transactions(from_date, to_date, supplier_filter):
    condition = "posting_date BETWEEN %(from_date)s AND %(to_date)s AND party_type = 'Supplier'"
    params = {"from_date": from_date, "to_date": to_date}

    if supplier_filter:
        condition += " AND party = %(supplier)s"
        params["supplier"] = supplier_filter

    result = frappe.db.sql(f"""
        SELECT party, voucher_type, voucher_subtype, voucher_no, posting_date,
               SUM(debit) AS debit, SUM(credit) AS credit
        FROM `tabGL Entry`
        WHERE {condition}
        GROUP BY voucher_type, voucher_no
        HAVING SUM(debit) != SUM(credit)
        ORDER BY party, posting_date
    """, params, as_dict=1)

    enriched = []
    for gle in result:
        details = ""
        p_name = ""

        if gle["voucher_type"] == "Purchase Invoice":
            pi = frappe.db.get_value(
                "Purchase Invoice", gle["voucher_no"],
                ["docstatus", "is_return", "bill_no", "remarks", "supplier_name"],
                as_dict=True
            )
            if not pi or pi.docstatus == 2:
                continue
            if pi.bill_no:
                details = f"Supp.Inv: {pi.bill_no}"
            elif pi.remarks:
                details = f"Remarks: {pi.remarks}"
            p_name = pi.supplier_name

        elif gle["voucher_type"] == "Payment Entry":
            pe = frappe.db.get_value(
                "Payment Entry", gle["voucher_no"],
                ["docstatus", "reference_no", "remarks", "party_name"],
                as_dict=True
            )
            if not pe or pe.docstatus == 2:
                continue
            if pe.reference_no:
                details = f"Ref: {pe.reference_no}"
            elif pe.remarks:
                details = f"Remarks: {pe.remarks}"
            p_name = pe.party_name

        elif gle["voucher_type"] == "Journal Entry":
            je = frappe.db.get_value("Journal Entry", gle["voucher_no"], "user_remark")
            if je:
                details = f"Remarks: {je}"

        gle["party_name"] = p_name
        gle["doc_ref"] = details or ""
        enriched.append(gle)

    # Group by supplier
    grouped = {}
    for row in enriched:
        grouped.setdefault(row.party, []).append(row)

    return grouped

def get_columns():
    return [
        {"label": "Date", "fieldname": "date", "fieldtype": "Date", "width": 120},
        {"label": "Supplier", "fieldname": "supplier", "fieldtype": "Link", "options": "Supplier", "width": 150},
        {"label": "Supplier Name", "fieldname": "supplier_name", "fieldtype": "Data", "width": 200},
        {"label": "Voucher Subtype", "fieldname": "voucher_subtype", "fieldtype": "Data", "width": 120},
        {"label": "Voucher No", "fieldname": "voucher_no", "fieldtype": "Dynamic Link", "options": "voucher_type", "width": 150},
        {"label": "Doc Ref", "fieldname": "doc_ref", "fieldtype": "Data", "width": 200},
        {"label": "Opening", "fieldname": "opening_balance", "fieldtype": "Currency", "width": 120},
        {"label": "Debit", "fieldname": "debit", "fieldtype": "Currency", "width": 120},
        {"label": "Credit", "fieldname": "credit", "fieldtype": "Currency", "width": 120},
        {"label": "Balance", "fieldname": "balance", "fieldtype": "Currency", "width": 120},
    ]
